### ISSUE TITLE
feature: document development plan to integrate pluggable transport 

### ISSUE THREAD

**Issue by nirrozenbaum** [kubestellar/kubestellar#1692]
### Feature Description

adding pluggable transport is split into multiple PRs and issues. 
a documented development plan is required for the understanding of the steps

### Proposed Solution

1. @vMaroon implemented the *minimal* set of changes to the placement controller to modify it such that it produces `PlacementDecision`. please pay attention that with this PR, placement controller is still doing the whole process of wrapping objects and put them into mailbox namespaces, but in parallel it also produces `PlacementDecision` objects.
this is done in the following issue - #1681 - **DONE** 

2. we should make sure that status-addon work correctly when wrapping multiple objects.
@andreyod should take care for that in the following issue - #1724  - **DONE**
this might be blocked by issue - #1605. we need to rehome status addon under the KS organization - rehome is **DONE**.

3. need to create docker image from the ocm-transport-plugin repo in order to be able to run containerized transport controller (based on ocm). in CICD we might consider build an executable from ocm-transport-plugin repo and use it (doesn't necessarily have to be containerized in CICD) - PR #1780 - **DONE**.

4. next, we should make sure the automated e2e tests + documentation are using the transport controller (using ocm plugin). transport controller watches `PlacementDecision` and creates wrapped objects from them. we need to verify e2e tests are passing PR #1731. 
for this we might need help with the e2e tests from @andreyod and @effi-ofer. - **DONE**

5. introduce a new PR to remove from placement controller the code that handles ocm specifics. placement controller should complete his work upon the creation/update of `PlacementDecision` - PR #1730 . 
Along with this PR, we need to make sure the architecture documentation is up to date in PR #1686. 
 

### Want to contribute?

- [ ] I would like to work on this issue.

**Comment by nirrozenbaum** [comment]
/cc @MikeSpreitzer 
/cc @ezrasilvera 
/cc @pdettori 

please see the above plan and add your comments if there are any.

please also keep in mind that the above is the *minimal* set of changes to make pluggable transport e2e work correctly.
other than the above, there are additional different issues to be implemented and improve the placement code.

**Comment by effi-ofer** [comment]
There are existing e2e tests that are script based. I suggest you wait till next week incase we switch to ginkgo.  In other words, there is a dependency on issue #1612 which we expect to resolve by the middle of next week.

**Comment by MikeSpreitzer** [comment]
I thought we just agreed to keep those two bash-based e2e scripts.

**Comment by MikeSpreitzer** [comment]
(A) The opening comment seems to outline a plan for upcoming changes. It would be good to recapitulate the starting point. What about the transport controller? I do not see anything that calls https://github.com/kubestellar/kubestellar/blob/ad2f86d60cf5b27a88a16839d419d152b346ff5a/cmd/transport/base-controller.go#L52 .

(B) If we take step 3 without step 4, do we have a working system? Wouldn't it be trying to transport two ways?

**Comment by nirrozenbaum** [comment]
> (A) The opening comment seems to outline a plan for upcoming changes. It would be good to recapitulate the starting point. What about the transport controller? I do not see anything that calls
> 
> https://github.com/kubestellar/kubestellar/blob/ad2f86d60cf5b27a88a16839d419d152b346ff5a/cmd/transport/base-controller.go#L52
> 

right. this is an "abstract" controller. the plugin specific implementation has to initialize the controller with the specific implementation of the transport interface.
I create the ocm plugin repo, but need to set all open-source requirements there.
In the meantime I just added basic makefile that does a simple build that outputs ocm based transport controller executable.
need of course to build docker image to run it in container.
maybe that's worth another issue in the development plan. 
until open source requirements are fulfilled, repo is private.
you can see it here:
https://github.com/kubestellar/ocm-transport-plugin/blob/main/cmd/main.go

> (B) If we take step 3 without step 4, do we have a working system? Wouldn't it be trying to transport two ways?

yes. e2e should still be passing. there is no harm in distributing the same object twice in two different manifest work objects.
so we should have a working system after step 3.
of course we don't want to distribute the same object twice and we want to disable the previous one.
so we should do step 4 and still be left with a working system.

**Comment by vMaroon** [comment]
This should also be done #1708. Trivial but the current #1681 is already huge.

**Comment by MikeSpreitzer** [comment]
@nirrozenbaum : the opening comment should have the full plan, including the additions discussed above.

**Comment by nirrozenbaum** [comment]
right.
@vMaroon can you please add items that were added to the plan as a result of #1681 review?
I'll add my items as well.

**Comment by nirrozenbaum** [comment]
additional items on our list that are not mandatory but are expected to be implemented soon:
- #1702 
- #1708

**Comment by vMaroon** [comment]
non-blocking items that were birthed in #1681:
- #1703
- #1716
- #1699

**Comment by nirrozenbaum** [comment]
closing this issue as done.
the development plan was documented and followed by the team.
the last open issue is to disable old distribution code, which is handled in PR #1765
